src/self-managed-debug/src/main.rs

```diff
@@ -0,0 +1,722 @@
+                    namespace, e
+                );
+                None
+            }
+        };
+
+        let _ = match dump_k8s_get_all(client.clone(), &namespace).await {
+            Ok(file_name) => Some(file_name),
+            Err(e) => {
+                eprintln!(
+                    "Failed to write k8s get all for namespace {}: {}",
+                    namespace, e
+                );
+                None
+            }
+        };
+    }
+
+    // TODO: Compress files to ZIP
+    Ok(())
+}
+
+/**
+ * Creates a k8s client given a context. If no context is provided, the default context is used.
+ */
+async fn create_k8s_client(context: Option<String>) -> Result<Client, anyhow::Error> {
+    let kubeconfig_options = KubeConfigOptions {
+        context,
+        ..Default::default()
+    };
+
+    let kubeconfig = Config::from_kubeconfig(&kubeconfig_options).await?;
+
+    let client = Client::try_from(kubeconfig)?;
+
+    Ok(client)
+}
+
+/**
+ * Write k8s pod logs to a file per pod as mz-pod-logs.<namespace>.<pod-name>.log.
+ * Returns a list of file names on success.
+ */
+async fn dump_k8s_pod_logs(
+    client: Client,
+    namespace: &String,
+) -> Result<Vec<String>, anyhow::Error> {
+    let mut file_names = Vec::new();
+
+    let pods: Api<Pod> = Api::<Pod>::namespaced(client.clone(), namespace);
+
+    let pod_list = pods.list(&ListParams::default()).await?;
+
+    for pod in &pod_list.items {
+        if let Err(e) = async {
+            let pod_name = pod.metadata.name.clone().unwrap_or_default();
+            let file_name = format!("mz-pod-logs.{}.{}.log", namespace, pod_name);
+            let mut file = File::create(&file_name)?;
+
+            let logs = match pods
+                .logs(
+                    &pod_name,
+                    &LogParams {
+                        previous: true,
+                        timestamps: true,
+                        ..Default::default()
+                    },
+                )
+                .await
+            {
+                Ok(logs) => logs,
+                Err(_) => {
+                    // If we get a bad request error, try without the previous flag.
+                    pods.logs(
+                        &pod_name,
+                        &LogParams {
+                            timestamps: true,
+                            ..Default::default()
+                        },
+                    )
+                    .await?
+                }
+            };
+
+            for line in logs.lines() {
+                writeln!(file, "{}", line)?;
+            }
+
+            println!("Finished exporting logs for {}", pod_name);
+            file_names.push(file_name);
+            Ok::<(), anyhow::Error>(())
+        }
+        .await
+        {
+            let pod_name = pod.metadata.name.clone().unwrap_or_default();
+            eprintln!("Failed to process pod {}: {}", pod_name, e);
+        }
```

Comment:
This might not do what you want: the `?` operator returns from the enclosing function, not the block.

Reply:
Ah but if we're referring to the usage of `?` in lines 132-168, wouldn't it return from the block given it's an async block?